docs/rfcs/0011-customizable-store-dispatcher.md

```diff
@@ -0,0 +1,71 @@
+
+There are some advantages to this change:
+
+* Simplicity for main-thread `Store`s.
+* Unit testing can now occur on the test framework's thread.
+* Fewer resources needed for context shifting between threads.
+
+## Guide-level explanation
+
+There are two possible ways to allow for this behaviour during the store creation:
+
+1. Pass in a `CoroutineDispatcher`.
+2. Configurable set of pre-defined options to choose.
+
+Passing in a `CoroutineDispatcher` has the flexibility to allow any preferred dispatcher to be used (e.g. `Dispatchers.IO` or Android Component's `Dispatchers.Cached`),  however that could lead to complications on choosing the wrong dispatcher. Ensuring that testing is also done with the same dispatcher configuration can also be difficult with this pattern.
+
+With a predefined set of options, we have a smaller surface area of changes that we want to maintain, possibly starting off with either the main dispatcher or the current single threaded dispatcher. We can continue to use our `StoreThreadFactory` and disable it when we switch to the `Main` dispatcher.
+
+Below are examples of what either surface would look like to a consumer of an "AppStore":
+
+```kotlin
+// Pass in a `CoroutineDispatcher`.
+class AppStore(
+    initialState: AppState = AppState(),
+) : Store<AppState, AppAction>(
+    initialState = initialState,
+    reducer = AppStoreReducer::reduce,
+    dispatcher = Dispatchers.Main // choose a preferable dispatcher.
+)
+
+// Configurable set of pre-defined options to choose.
+class AppStore(
+    initialState: AppState = AppState(),
+) : Store<AppState, AppAction>(
+    initialState = initialState,
+    reducer = AppStoreReducer::reduce,
+    runOnMain = true, // a boolean or enum value to signify the dispatcher used.
+)
+```
+
+Another advantage of this change is that `Store`s would be able to run on the test framework's default dispatcher. This reduces the number of intermittent test failures we have seen where the `Store` is still processing actions async to an assertion. Fewer usages of `store.waitUntilIdle()` are required after every `dispatch` is called in a test, as a result.
+
+## Drawbacks
+
+* Mistakenly doing work on the main thread - we could end up performing large amounts of work on the main thread unintentionally if we are not careful. This could be because of a large number of small tasks, a single large task, a blocking task, or a combination. This could be mitigated by keeping the `Store` dispatcher as the default, and catching performance regressions in the specific features they are used in.
```

Comment:
Is there something we could do to mitigate this risk by introducing restrictions on `Middleware`? They are the most likely culprit for long-running operations since they handle side-effects.

Reply:
Aside from the mitigation already mentioned, I can't see any obvious ways to do this programatically. Each application of the `Store` would have to know what work they are performing and how large it is to avoid this.

So it would be up to them to know what work is considered non-performant. In Fenix, there is process to can performance regressions and that has been used to find bugs where too many actions were being dispatched (e.g. download progress was [too noisy](https://github.com/mozilla-mobile/android-components/pull/8197/files#diff-51a4f62b18b097b25a5482b31e10be7850f9a23d1c097fecaa7fb2d4f6674bc1R635-R641).

Reply:
This is not a problem we need to solve at the component-level, or at least not prematurely. @rahulsainani has recently added some helpful code to pass [in the scope used to create the store for middlewares](https://searchfox.org/mozilla-mobile/source/firefox-android/fenix/app/src/main/java/org/mozilla/fenix/components/StoreProvider.kt#61-81), so this might be an area we can consider improving there first and we can go from there.

Reply:
Rahul's patch is very useful for real usecases, but isn't usable in unit tests where we construct Stores without `StoreProvider.get()`. This proposal would allow us to more easily share the test scope between middleware and Stores, but I still think those could be automatically shared in the future. Agreed that this is tangential in regards to this proposal though